R/bundlePackage.R

```diff
@@ -103,75 +102,89 @@ findPackageRepoAndSource <- function(records, repos) {
+  rows <- lapply(seq_len(nrow(records)), function(i) {
+    standardizePackageRepoAndSource(records[i, ], availablePackages, repos = repos)
+  })
+  rows <- lapply(rows, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(rows, c("Source", "Repository"))
+}
+
+standardizePackageRepoAndSource <- function(record, availablePackages, repos = character()) {
+
+  pkg <- record$Package
+  source <- record$Source
 
-    pkg <- records[i, "Package"]
-    source <- records[i, "Source"]
+  # source types are defined by packrat:
+  # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
+    repository <- NA
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    source <- NA
     repository <- NA
-    # capture Bioconcutor repository
-    if (identical(source, "Bioconductor")) {
-      if (pkg %in% biocPackages) {
-        repository <- biocPackages[pkg, "Repository"]
-      }
-    } else if (isSCMSource(source)) {
-      # leave source+SCM packages alone.
-    } else if (pkg %in% rownames(repo.packages)) {
-      # capture CRAN-like repository
-
-      # Find this package in the set of available packages then use its
-      # contrib.url to map back to the configured repositories.
-      package.contrib <- repo.packages[pkg, "Repository"]
-      package.repo.index <- vapply(repo.lookup$contrib.url,
-                                   function(url) grepl(url, package.contrib, fixed = TRUE), logical(1))
-      package.repo <- repo.lookup[package.repo.index, ][1, ]
-      # If the incoming package comes from CRAN, keep the CRAN name in place
-      # even if that means using a different name than the repos list.
-      #
-      # The "cran" source is a well-known location for shinyapps.io.
-      #
-      # shinyapps.io isn't going to use the manifest-provided CRAN URL,
-      # but other consumers (Connect) will.
-      if (tolower(source) != "cran") {
-        source <- package.repo$name
-      }
-      repository <- package.repo$url
+  } else if (source == "CustomCRANLikeRepository") {
+    # Packrat guessed that this package was installed from CRAN, but
+    # we need to check that it's not actually a development version
+    if (is_dev_version(record, availablePackages)) {
+      source <- NA
+      repository <- NA
+    } else {
+      repository <- findRepoUrl(pkg, availablePackages)
     }
-    # validatePackageSource will emit a warning for packages with NA repository.
-    data.frame(Source = source, Repository = repository, stringsAsFactors = FALSE)
-  })
-  do.call("rbind", tmp)
+  } else if (source %in% c("CRAN", "Bioconductor")) {
+    # Installed from standard repository
+    repository <- findRepoUrl(pkg, availablePackages)
+  } else {
+    # Installed from custom repository: source is currently the repo url,
+    # but we need it to be the repo name (as defined in `getOptions("repos")`).
```

Comment:
Am I misunderstanding the Packrat choice in `pkg.R`? I was thinking that the incoming source was the value of `packageDescription()$Repository`, not the repository URL.

Reply:
```R
install.packages('cffr', repos = c('https://ropensci.r-universe.dev', 'https://cloud.r-project.org'))
packageDescription("cffr")$Repository
#> [1] "https://ropensci.r-universe.dev"
```

`packageDescription()$Repository` is the repository URL?

Reply:
Oh but

```R
install.packages("filelock", repos = "http://rspm.infra/all/__linux__/bionic/latest")
packageDescription("filelock")$Repository
#> [1] "RSPM"
```

```R
install.packages("filelock", repos = "http://cloud.r-project.org")
packageDescription("filelock")$Repository
#> [1] "CRAN"
```

Reply:
The RSPM/CRAN cases were the ones I've commonly seen. I've also seen cases where the `Repository` fields was not populated. I had not seen the ropensci case with a URL as the `Repository` field.